### PR TITLE
Fix misleading notion that Flatten op is not supported for opset 10

### PR DESCRIPTION
**Pull request by hariharans29**
For the sake of consistency with all other declared kernels, remove the end version for the kernel declaration (Takeaway from #1088)

### CODE REVIEW

**Comment by snnn** [comment]
I don't understand. Why do you change this?
The latest version of this OP is 9. Do we support it or not?

**Comment by skottmckay** [comment]
> I don't understand. Why do you change this?
> The latest version of this OP is 9. Do we support it or not?

Opset 9 just added some more types to Flatten, so we can just use the one kernel for opset 1 on, and don't need to set an end version at this point.

**Comment by snnn** [comment]
> > I don't understand. Why do you change this?
> > The latest version of this OP is 9. Do we support it or not?
> 
> Opset 9 just added some more types to Flatten, so we can just use the one kernel for opset 1 on, and don't need to set an end version at this point.

[1,9] means 1 to  9, inclusive, but doesn't include 10.
[1] means just version 1, no others.
Am I wrong?

**Comment by hariharans29** [comment]
> > > I don't understand. Why do you change this?
> > > The latest version of this OP is 9. Do we support it or not?
> > 
> > 
> > Opset 9 just added some more types to Flatten, so we can just use the one kernel for opset 1 on, and don't need to set an end version at this point.
> 
> [1,9] means 1 to 9, inclusive, but doesn't include 10.
> [1] means just version 1, no others.
> Am I wrong?

@snnn - My understanding is [1] means 1 *onwards* and not just 1.

**Comment by skottmckay** [comment]
> @snnn - My understanding is [1] means 1 _onwards_ and not just 1.

I believe this is correct. If there's only one value, it's from that value on. If there are two values it's the start to (inclusive) end. 

So whilst [1, 9] will match both the opset 1 and opset 9 version of Flatten it is inconsistent (usually the end version is the last one prior to a change in the op) and confusing. If there's a new version of Flatten in opset 11 you'd have one defined as [1, 9] and another as [11], making it unclear what happens with 10 unless you understand all the intricacies of the matching. 

Having [1,8] and [9] is how we'd normally do it, and if a new version was added in opset 11 you'd change the [9] to [9, 10] and add an [11], giving you [1,8], [9,10] and [11] which is far more intuitive.

**Comment by hariharans29** [comment]
> Can we change to define a version for opset 1 - 9 and one for opset 10 onwards, both using the same implementation?
> 
> Is it possible to also update the CUDA version of Flatten the same way?

Hi @skottmckay - Yes I will do that.

**Comment by hariharans29** [comment]
> Can we change to define a version for opset 1 - 9 and one for opset 10 onwards, both using the same implementation?
> 
> Is it possible to also update the CUDA version of Flatten the same way?

Hi @skottmckay,

Do you think it would be better if we defined one kernel for opset 1 to 8 and one kernel from opset 9 onwards (both pointing to the same implementation of course) given that Flatten underwent versioning change for opset 9 ?

**Comment by hariharans29** [comment]
CC: @yuslepukhin

**Comment by skottmckay** [comment]
> Do you think it would be better if we defined one kernel for opset 1 to 8 and one kernel from opset 9 onwards (both pointing to the same implementation of course) given that Flatten underwent versioning change for opset 9 ?

Absolutely. Sorry - I thought it changed in 10 not 9. 1 - 8 is correct.

File: onnxruntime/core/providers/cpu/nn/flatten.cc [hunk 1]
```diff
@@ -7,6 +7,14 @@ namespace onnxruntime {
 ONNX_CPU_OPERATOR_VERSIONED_KERNEL(
     Flatten,
     1,
+	8,
+    KernelDefBuilder()
```
Comment (skottmckay) [inline onnxruntime/core/providers/cpu/nn/flatten.cc]: nit: tab

File: onnxruntime/core/providers/cuda/cuda_execution_provider.cc [hunk 2]
```diff
@@ -537,7 +538,7 @@ static void RegisterCudaKernels(KernelRegistry& kernel_registry) {
       BuildKernelCreateInfo<ONNX_OPERATOR_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 1, MemcpyToHost)>,
       BuildKernelCreateInfo<ONNX_OPERATOR_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 4, Concat)>,
       BuildKernelCreateInfo<ONNX_OPERATOR_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 1, Unsqueeze)>,
-      BuildKernelCreateInfo<ONNX_OPERATOR_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 1, Flatten)>,
+      BuildKernelCreateInfo<ONNX_OPERATOR_VERSIONED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 1, 8, Flatten)>,
       BuildKernelCreateInfo<ONNX_OPERATOR_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 1, Squeeze)>,
```
Comment (skottmckay) [inline onnxruntime/core/providers/cuda/cuda_execution_provider.cc]: Should onnxruntime\core\providers\cuda\tensor\flatten.cc get a similar change to what the CPU version had?

**Pull request merged**